api/src/OpenTelemetry/Internal/Logging/Types.hs

```diff
@@ -44,7 +46,10 @@ data Logger = Logger
 {- | This is a data type that can represent logs from various sources: application log files, machine generated events, system logs, etc. [Specification outlined here.](https://opentelemetry.io/docs/specs/otel/logs/data-model/)
 Existing log formats can be unambiguously mapped to this data type. Reverse mapping from this data type is also possible to the extent that the target log format has equivalent capabilities.
 -}
-data LogRecord body = LogRecord
+data LogRecord a = LogRecord (IORef (ImmutableLogRecord a))
+
+
+data ImmutableLogRecord body = ImmutableLogRecord
```

Comment:
I am curious about the reason Ian used an IORef instead of something like the State monad. I'll check with him about that when he gets back.

Reply:
As a thought exercise, consider adding logging to an `IO` function. You have a `Logger`, you create a `LogRecord body` value.

You do some IO and want to add some attributes to the `LogRecord body`, then emit the log. Usage of the state monad doesn't work for that unless you want to force people to rewrite their monad stack, which we don't.

With regards to why we use an IORef, it's because we're more or less told that we have to do it that way: https://opentelemetry.io/docs/specs/otel/logs/sdk/#additional-logrecord-interfaces

In that regard, we probably want to update our naming here to reflect the naming outlined in that section.

Reply:
That makes sense. In regards to naming, `Span`s also have "readable" and "read/write" versions defined in the spec, but `hs-opentelemetry`only implements the read/write spans. The spec says that readable log records can be implemented as read/write log records instead. Essentially, since `LogRecord` is both a `ReadableLogRecord` and `ReadWriteLogRecord`, I thought it would be simplest to call it `LogRecord`. It would be possible to do both, but `ReadableLogRecord` would have to be somethings like

```haskell
module Readable (
  ReadableLogRecord, -- not exporting constructor
  readReadableLogRecord
) where

-- It needs to be an IORef because it has to contain all changes made to the LogRecord
data ReadableLogRecord body = ReadableLogRecord (IORef (ImmutableLogRecord body))

readReadableLogRecord :: ReadableLogRecord body -> ImmutableLogRecord body
```

Reply:
I ended up writing a typeclass for `ReadableLogRecord`s and `ReadWriteLogRecord`s.